### PR TITLE
doc: Add documentation for JSONRPC class

### PR DESCRIPTION
**Pull request by IcedQuinn**
should be less cryptic now

### CODE REVIEW

File: doc/classes/JSONRPC.xml [hunk 1]
```diff
@@ -1,8 +1,11 @@
 <?xml version="1.0" encoding="UTF-8" ?>
 <class name="JSONRPC" inherits="Object" version="4.0">
 	<brief_description>
+	A helper to handle dictionaries which look like JSONRPC documents.
```
Comment (Calinou) [inline doc/classes/JSONRPC.xml:4]:
```suggestion
		A helper to handle dictionaries which look like JSONRPC documents.
```

File: doc/classes/JSONRPC.xml [hunk 2]
```diff
@@ -1,8 +1,11 @@
...
 	</brief_description>
 	<description>
+    	JSON-RPC is a standard which wraps a method call in a JSON object. The object has a particular structure and identifies which method is called, the parameters to that function, and carries an ID to keep track of responses. This class implements that standard on top of Dictionaries; you will have to convert between a Dictionary and JSON with other functions.
+        See: https://www.jsonrpc.org/specification
```
Comment (Calinou) [inline doc/classes/JSONRPC.xml:8]:
```suggestion
		[url=https://www.jsonrpc.org/]JSON-RPC[/url] is a standard which wraps a method call in a [JSON] object. The object has a particular structure and identifies which method is called, the parameters to that function, and carries an ID to keep track of responses. This class implements that standard on top of [Dictionary]; you will have to convert between a [Dictionary] and [JSON] with other functions.
```

**Comment by Calinou** [comment]
There are several issues you need to fix:

- Only tab indentation should be used.
- Individual arguments cannot have descriptions, so you need to describe each argument in the method description instead (use `[code]argument_name[/code]` to refer to arguments in the description).

**Review by Calinou — DISMISSED** [review]

File: doc/classes/JSONRPC.xml [hunk 3]
```diff
@@ -11,42 +13,55 @@
 			<return type="Dictionary">
 			</return>
 			<argument index="0" name="method" type="String">
+    			Name of the method being called.
```
Comment (akien-mga) [inline doc/classes/JSONRPC.xml:16]: There are indentation issues where spaces and tabs are mixed. All indentation should be done with tabs.

**Comment by IcedQuinn** [comment]
is this still missing something?

File: doc/classes/JSONRPC.xml [hunk 4]
```diff
@@ -49,16 +61,24 @@
 			<argument index="2" name="id" type="Variant" default="null">
 			</argument>
 			<description>
+			Creates a response which indicates a previous reply has failed in some way.
+			[code]The error code corresponding to what kind of error this is. See the ErrorCode constants.[/code]
+			[code]A custom message about this error.[/code]
+			[code]The request this error is a response to.[/code]
```
Comment (akien-mga) [inline doc/classes/JSONRPC.xml:67]: I guess this was meant to be:
`[code]name[/code] The error code corresponding to what kind of error this is. See the ErrorCode constants.`
etc.?

File: doc/classes/JSONRPC.xml [hunk 5]
```diff
@@ -49,16 +61,24 @@
...
 			</return>
 			<argument index="0" name="action" type="Variant">
+				
 			</argument>
 			<argument index="1" name="recurse" type="bool" default="false">
 			</argument>
 			<description>
+				Given a Dictionary which takes the form of a JSON-RPC request: unpack the request and run it. Methods are resolved by looking at the field called "method" and looking for an equivalently named function in the JSONRPC object. If one is found that method is called.
+				To add new supported methods extend the JSONRPC class and call [code]process_action[/code] on your subclass.
+				[code]The action to be run, as a Dictionary in the form of a JSON-RPC request or notification.[/code]
```
Comment (akien-mga) [inline doc/classes/JSONRPC.xml:81]: Same here.

File: doc/classes/JSONRPC.xml [hunk 6]
```diff
@@ -15,6 +17,9 @@
 			<argument index="1" name="params" type="Variant">
 			</argument>
 			<description>
+				Returns a dictionary in the form of a JSON-RPC notification. Notifications are one-shot messages which do not expect a response.
+				- [code]method[/code] Name of the method being called.
+				- [code]params[/code] An array or dictionary of parameters being passed to the method.
```
Comment (akien-mga) [inline doc/classes/JSONRPC.xml:22]: I would add a colon after the name of the key:
```suggestion
				- [code]method[/code]: Name of the method being called.
				- [code]params[/code]: An array or dictionary of parameters being passed to the method.
```

Same for all other descriptions with details on the keys.

**Comment by akien-mga** [comment]
Looks great overall, just found a few remaining issues to address.

Then the commits in this PR need to be squashed into one so that the PR is ready for merging. See [PR workflow](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase) for instructions.

**Comment by IcedQuinn** [comment]
ok.

**Comment by akien-mga** [comment]
Thanks. I force pushed an update to rebase and solve merge conflits, and fix remaining style issues.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

**Comment by akien-mga** [comment]
Cherry-picked for 3.3.3.

**Pull request merged**